File: backend/app/services/google_oauth.py
from __future__ import annotations

from dataclasses import dataclass

import httpx

from app.config import get_settings

GOOGLE_TOKEN_URL = "https://oauth2.googleapis.com/token"
GOOGLE_USERINFO_URL = "https://www.googleapis.com/oauth2/v2/userinfo"


@dataclass(frozen=True)
class GoogleUser:
    google_id: str
    email: str
    name: str
    picture: str | None


async def exchange_code_for_user(code: str) -> GoogleUser:
    """Exchange an authorization code for Google user info."""
    settings = get_settings()
    if not settings.google_client_id or not settings.google_client_secret:
        raise ValueError("Google OAuth is not configured.")

    async with httpx.AsyncClient(timeout=15) as client:
        token_response = await client.post(
            GOOGLE_TOKEN_URL,
            data={
                "code": code,
                "client_id": settings.google_client_id,
                "client_secret": settings.google_client_secret,
                "redirect_uri": settings.google_redirect_uri,
                "grant_type": "authorization_code",
            },
        )
        token_response.raise_for_status()
        tokens = token_response.json()

        userinfo_response = await client.get(
            GOOGLE_USERINFO_URL,
            headers={"Authorization": f"Bearer {tokens['access_token']}"},
        )
        userinfo_response.raise_for_status()
        info = userinfo_response.json()

    return GoogleUser(
        google_id=info["id"],
        email=info["email"],
        name=info.get("name", info["email"].split("@")[0]),
        picture=info.get("picture"),
    )

File: backend/internal/services/google_oauth.go
package services

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"triequest-backend/internal/config"
)

const (
	googleTokenURL    = "https://oauth2.googleapis.com/token"
	googleUserInfoURL = "https://www.googleapis.com/oauth2/v2/userinfo"
)

type GoogleUser struct {
	GoogleID      string
	Email         string
	Name          string
	Picture       *string
	EmailVerified bool
}

func ExchangeCodeForUser(code string, cfg *config.Settings) (*GoogleUser, error) {
	if cfg.GoogleClientID == "" || cfg.GoogleClientSecret == "" {
		return nil, fmt.Errorf("Google OAuth is not configured")
	}

	client := &http.Client{Timeout: 15 * time.Second}

	// Exchange code for tokens
	form := url.Values{}
	form.Set("code", code)
	form.Set("client_id", cfg.GoogleClientID)
	form.Set("client_secret", cfg.GoogleClientSecret)
	form.Set("redirect_uri", cfg.GoogleRedirectURI)
	form.Set("grant_type", "authorization_code")

	tokenResp, err := client.PostForm(googleTokenURL, form)
	if err != nil {
		return nil, fmt.Errorf("token exchange failed: %w", err)
	}
	defer tokenResp.Body.Close()

	if tokenResp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token exchange returned status %d", tokenResp.StatusCode)
	}

	var tokens struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.NewDecoder(tokenResp.Body).Decode(&tokens); err != nil {
		return nil, fmt.Errorf("failed to decode token response: %w", err)
	}

	// Fetch user info
	req, _ := http.NewRequest("GET", googleUserInfoURL, nil)
	req.Header.Set("Authorization", "Bearer "+tokens.AccessToken)
	infoResp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("user info request failed: %w", err)
	}
	defer infoResp.Body.Close()

	if infoResp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("user info returned status %d", infoResp.StatusCode)
	}

	var info struct {
		ID            string `json:"id"`
		Email         string `json:"email"`
		Name          string `json:"name"`
		Picture       string `json:"picture"`
		VerifiedEmail bool   `json:"verified_email"`
	}
	if err := json.NewDecoder(infoResp.Body).Decode(&info); err != nil {
		return nil, fmt.Errorf("failed to decode user info: %w", err)
	}

	name := info.Name
	if name == "" {
		name = strings.Split(info.Email, "@")[0]
	}

	var picture *string
	if info.Picture != "" {
		picture = &info.Picture
	}

	return &GoogleUser{
		GoogleID:      info.ID,
		Email:         info.Email,
		Name:          name,
		Picture:       picture,
		EmailVerified: info.VerifiedEmail,
	}, nil
}
